kroxylicious/src/main/java/io/kroxylicious/proxy/codec/CorrelationManager.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.kroxylicious.proxy.codec;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Manages correlation ids for a single connection (across the proxy) between a single client
+ * and a single broker.
+ */
+public class CorrelationManager {
+
+    // TODO use a specialized map
+    /* test */ final Map<Integer, Correlation> brokerRequests = new HashMap<>();
+
+    private final AtomicInteger upstreamIds;
+
+    public CorrelationManager() {
+        this(0);
+    }
+
+    public CorrelationManager(int initialCorrelationId) {
+        upstreamIds = new AtomicInteger(initialCorrelationId);
+    }
+
+    /**
+     * Allocate and return a correlation id for an outgoing request to the broker.
+     * @param apiKey The
+     * @param apiVersion
+     * @param downstreamCorrelationId
+     * @param decodeResponse
+     * @return
+     */
+    public int putBrokerRequest(short apiKey,
+                                short apiVersion,
+                                int downstreamCorrelationId,
+                                boolean hasResponse,
+                                boolean decodeResponse) {
+        // need to allocate an id and put in a map for quick lookup, along with the "tag"
+        int upstreamCorrelationId = upstreamIds.getAndIncrement();
+        if (hasResponse) {
+            this.brokerRequests.put(upstreamCorrelationId, new Correlation(apiKey, apiVersion, downstreamCorrelationId, decodeResponse));
+        }
+        return upstreamCorrelationId;
+    }
+
+    /**
+     * Find (and remove) the Correlation for an incoming response from the broker
+     * @param correlationId The (upstream) correlation id in the response.
+     * @return
+     */
+    public Correlation getBrokerCorrelation(int correlationId) {
+        // Set the tag on the response object
+        return brokerRequests.remove(correlationId);
+    }
+
+    /**
+     * A record for which responses should be decoded, together with their
+     * API key and version.
+     */
+    // TODO a perfect value type
+    public static class Correlation {
```

Comment:
This is a good record type indeed

Reply:
There's a possible optimization which would avoid us needing `Correlation` entirely. The idea would be to use the `RequestFrame` _as the `Correlation`_ (at the cost of needing an extra `int` for storing the upstream correlation id). i.e. The `CorrelationManager` would put the `RequestFrame` in the map, nulling out the request `header` and `body` (so they can be GCd).

TBH I've no idea how much allocation paranoia is appropriate; it feels like premature optimization at this point, but in combination with using a specialised hash map in `CorrelationManager` (rather than the `HashMap<Integer, Correlation>`) maybe it's (eventually) worth doing, because for opaque frames it would reduce the amount of allocation a lot (possibly 1 allocation/(request+response) would be possible).

Reply:
> This is a good record type indeed

Ok, but we still can't use Java 17, right, @franz1981?

Reply:
Just to be clear, I meant [`value class`](https://openjdk.org/jeps/8277163) in the Valhalla sense. But it's also a good Java 17 `record` type in the mean time.

Reply:
> We can try and see what fails

Probably a good idea to talk about the Java 17 move in tomorrow's call.

> I meant [value class](https://openjdk.org/jeps/8277163) in the Valhalla sense.

Oooh, you're way ahead of the curve ;) Was thinking the same for the correlation id btw, it was really itching me that it is a "raw" int.

Reply:
Added the Java 17 question to today's meeting agenda.